source/sonata_simulation.rst

```diff
@@ -297,7 +297,7 @@ Spike events are created from the cells indicated in a file and delivered to the
    Property                       Type       Requirement  Description
    ============================== ========== ============ ==========================================
    spike_file                     text       Mandatory    Indicates the location of the file with the spike info for injection.
-   source                         text       Optional     The node set to replay spikes from.
+   source                         text       Optional     The node set to replay spikes from, recommended in case of multiple populations. By default, replay spikes are applied to the cells in all the populations that match the ids in the "spike_file".
```

Comment:
If I'm reading this right, I would find this behavior pretty surprising - it basically means all populations would fire simulataneously, when I think the expected behavior would only ever be for a single population to file.  
Perhaps the requirement should be made mandatory, so this behavior is only invoked when it's explicitly 'All'?

Perhaps @seirios, @andrisecker or another scientist could chime in (can you suggest someone @MWolfR?)

Reply:
This parameter is optional in BlueConfig "StimulusInject" where it wasn't used a lot because we had mostly single population.
Now with the multi-population in SONATA, users need to set the replay "source" more and more often to replay correctly. 
Making it mandatory sounds good to avoid mistakes.

Reply:
If I understand correctly this is remnants of the `.dat` input spike format where there's only ids and timestamps defined, no populations. In the `.h5` format, the timestamps and ids are defined per population, in which case, I could imagine further specifying a node set being a hindrance. 

If it's either/or, I'd agree and make `source` mandatory. We could also have it mandatory only if the input file is `.dat` and optional for `.h5`, which, to me, seems like the best of both worlds. That would also be in anticipation of the possible drop of `.dat` file support.

Reply:
I think neurodamus still requires "source" from the config when replaying with SONATA spikes file. It checks the required population when reading the SONATA spikes file.  @ferdonline , could you please confirm?

Reply:
In the middle of the subtle details, I think the current situation is as follows: source is not mandatory and when is not provided it will be translated as the ALL target, i.e. all instantiated populations (for BlueConfigs this used to be the "main-circuit" population).
Given there don't seem to be technical limitations, I think it's up to us to define the rules, especially if that protects the user from bad behavior.
I like Joni's suggestion. With recent multi-population circuits `.dat` replay without setting a population would create a bad setup. Sonata replay IMO can (maybe should) stay as is, so all applicable in the file are replayed.

Reply:
Could we just not support `.dat` files? It's a legacy format that's not even part of SONATA, and I think it would clean things up a lot if it wasn't supported at all.

Reply:
There are many users replaying with `.dat` files. I don't know if they are ready to migrate. Is there a tool to convert `.dat` to `.h5`?

Reply:
Since they area going to have to update the simulation_config to add the `source`; maybe it's the best time to have to move to the SONATA style.

A tool to convert the .dat files shouldn't be too hard to create; I wrote the skeleton of one a while back: I could add it to `brainbuilder`; but I'd want confirmation that it results in the same results as neurodamus.  Maybe you could help with that?

Reply:
Sure, please tag me once you have the PR/MR. 
The only tricky part I see is the 1-based gids in `.dat` to 0-based gids in `.h5`.

Reply:
I used this snippet to convert `.dat` to `.h5` while testing the validator code:
```python
import h5py
import pandas as pd
import numpy as np

populations = ['pop_name_1', 'pop_name_2']
dat_file = 'path/to/spikes.dat'
output_file = 'path/to/spikes.h5'

spikes = pd.read_csv(dat_file, delimiter=r"\s+", skiprows=1, header=None, names=["t", "id"])

with h5py.File(output_file, 'w') as h5:
    h5_spikes = h5.create_group('spikes')
    for population in populations:
        h5_pop = h5_spikes.create_group(population)
        h5_pop.create_dataset('timestamps', data=spikes.t, dtype=np.float64)
        h5_pop.create_dataset('node_ids', data=spikes.id-1, dtype=np.uint64)
```

I don't know if it covers everything, tho

Reply:
I think we should involve scientists here to know how they create the replay files. Even if we have a tool to convert from `.dat` they still need a way to create these files by hand, and we know they don't necessarily know hdf5. And to avoid them indefinitely creating .dat and converting (with the quirks involved), shouldn't we support a human-readable format as part of Sonata?

Reply:
I include @jmbudd here because we discussed that before. He mentioned that the replay files are artificially generated spike trains but in future could be experimental data. Maybe he can give more ideas.

If we make a converter, I guess it should be accessible for both BBP users and external OBP users.

Reply:
Simple converter here: if you could try it, that would be great:

https://bbpgitlab.epfl.ch/nse/brainbuilder/-/merge_requests/74

Can be used like:
```
brainbuilder sonata spike-converter --input path/to/some.dat --output out.h5 --population "test"
```

Reply:
The other thing is that the .dat file doesn't have a spec that I know of, so that results in confusion & differences between implementations, which causes headaches.

Reply:
We should make a decision about this, I think.
I propose we drop the `.dat` altogether - people can either convert it using the tools provided, or make their own with h5py.  If we don't drop `.dat`, then there is ambiguity on what `source` means.

In addition, the spike report includes the population (https://sonata-extension.readthedocs.io/en/latest/sonata_report.html#spike-file), so then the `source` can be the unambiguously the original intention: a node set name that narrows down which nodes are used from the population.

Reply:
> I propose we drop the .dat altogether

I second the proposal.

Reply:
Hello all, Thank you for your comments. I think we all agree to drop the `.dat` files for replay. Before merging this PR and changing the libsonata API, I am thinking that we need :

 [ ] Deploy the `spike_converter` (https://github.com/BlueBrain/libsonatareport/pull/41 ) in the libsonata-report repo on bb5. @jorblancoa 
 [ ] Inform users to convert their current ".dat" files to ".h5" files with `spike_converter`and  announce the date to drop `.dat` files. @WeinaJi

Reply:
I generated a module for libsonata-report, so now one can do:
```
module load unstable libsonata-report
spikes_converter --help
reports_converter --help
```

Reply:
Thank you @jorblancoa 
I've pushed a commit that removes the `.dat` replay from the SONATA spec. Can someone review it? 
After merging this PR, @jorblancoa  or I will announce in the software-general channel about this change and the instruction for spikes_converter.

Reply:
@GianlucaFicarelli noted that the units value needs to be set to nA in some cases:
https://bbpteam.epfl.ch/project/issues/browse/NSETM-2286?focusedId=235286&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-235286

I tried it with the `AllCompartments.h5` in `proj12/NSE/blueetl/data/sim-campaign-bluepy/11567ba9-57fa-41ec-ac7b-d727ec4e65f3/0`

and indeed the units are mV; 
```
In [6]: list(h5['report/All/data'].attrs.items())
Out[6]: [('units', 'mV')]
```

Looking at the Brion reader, it looks like there is a `D_UNIT`, could that not be used?
https://bbpteam.epfl.ch/source/xref/github/BlueBrain/Brion/brion/plugin/compartmentReportBinary.cpp?r=13c1e6df4b951242077e2937b85f3667c3d3b735&fi=MAPPING_NAME#70

Reply:
I could have a look at that.
I created a ticket with low priority, it it is needed let me know and I could work on that sooner:
https://bbpteam.epfl.ch/project/issues/browse/REP-87

Reply:
The fix has been deployed, now the units should be taken into account.
@mgeplf